### DISCUSSION TITLE
TOTP Registering device fails

### DISCUSSION THREAD

**Discussion by cganglin** [authelia/authelia#7301] [Q&A] [answered]
Would appreciate some pointers on resolving an issue i have.  Did a clean install of Authelia (running on proxmox VM) - not Docker.  I can log into the Authelia GUI but when i try Registering Device it fails - no email, no QR Code just sends me to a settings page for two form authenticator.  In the authelia.log I see the following:
Error occurred retrieving TOTP configuration for usertime="2024-05-12T05:53:09-04:00" level=error msg="Error occurred retrieving TOTP configuration for user 'username-removed': error occurred retrieving the configuration from the storage backend"
  
Any hints in resolving this?  Where would I find out where Authelia is looking for user TOTP configuration?    Thanks in advance.

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

It seems like the issue you're encountering is related to Authelia's inability to retrieve the Time-based One-Time Password (TOTP) configuration for your user from the storage backend. This could be due to several reasons such as misconfiguration of the storage backend, issues with the database connection, or permissions.

Here are a few steps you can take to troubleshoot and potentially resolve the issue:

1. **Check Database Connection**: Ensure that Authelia is correctly configured to connect to your database. This includes verifying the database host, port, username, password, and database name in your `configuration.yml` file.

2. **Database Permissions**: Make sure that the database user has the necessary permissions to read from the database that stores the TOTP configurations.

3. **Review Authelia Configuration**: Look at the `totp` section in your `configuration.yml` file. Ensure that it is correctly set up as per the Authelia documentation. You should check parameters like `issuer`, `algorithm`, `digits`, `period`, `skew`, and `secret_size`.

4. **Logs and Errors**: Since the error message mentioned issues with the storage backend, check the logs for any additional errors that might indicate what is wrong with the storage backend configuration.

5. **Encryption Keys**: If your TOTP secrets are encrypted in the database (which is recommended), ensure that Authelia is configured with the correct encryption keys to decrypt this information.

6. **Authelia Version**: Ensure that you are running a compatible version of Authelia that supports all configured features and check for any known issues with your version regarding TOTP.

If after checking these areas the issue persists, you might want to consult the Authelia community or support, providing them with the relevant configuration details (ensuring you redact sensitive information) and log outputs for more targeted assistance.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment]
Sorry for not getting back sooner. Did you manage to solve this?

**Comment by mouchabeurre** [comment]
Hi @james-d-elliott 
I think I have the same issue on my recent installation:
1FA working fine, but logs the following error when accessing the 2FA section in the user settings:
```
time="2024-07-19T09:40:14Z" level=error msg="Error occurred retrieving TOTP configuration for user 'username': error occurred retrieving the configuration from the storage backend" error="no TOTP configuration for user" method=GET path=/api/secondfactor/totp
``` 
authelia version v4.38.9
totp section in config:
```yaml
totp:
  disable: false
  issuer: 'authelia mydomain'
``` 
Should I open an issue to share more details?

**Reply by james-d-elliott** [reply]
A log message is not specifically a problem. Please elaborate on the issue you're having.

**Reply by mouchabeurre** [reply]
Yes, sorry!
My actual problem is that I can log into authelia's web ui (first ever connection via LDAP authentication for this user, so 1FA only), but when I try to add a device for 2FA by navigating to the "Two-Factor Authentication" section in the account settings, I get a red notification / snackbar saying "There was an issue retrieving the WebAuthn Credentials" and the settings page body remains empty, so no possible enrollment of a new device.
The payload of the 404 HTTP response following my navigation contains `{"status":"KO","message":"Could not find TOTP Configuration for user."}`, which I think is related to the log error I shared above.

**Reply by james-d-elliott** [reply]
What does the screen look like when the snackbar appears?

**Reply by mouchabeurre** [reply]
Screenshot right after I clicked "Two-Factor Authentication":
![Screenshot 2024-07-19 at 12-25-05 Settings - Authelia](https://github.com/user-attachments/assets/27d1ac0b-4c63-4edb-9cc0-fb514619c5da)

**Reply by james-d-elliott** [reply]
What happens when you refresh the page? Does refreshing add logs? What does the console and network tab look like in the browser dev tools?

**Reply by mouchabeurre** [reply]
- Refreshing adds the same error log as above (with some request timeout errors ; probably unrelated? seems to happen on page load)
```
time="2024-07-19T12:57:57Z" level=debug msg="Loaded Configuration Sources" files="[/config/configuration.yml]" filters="[template]"
time="2024-07-19T12:57:57Z" level=debug msg="Logging Initialized" fields.level=debug file= format= keep_stdout=false
time="2024-07-19T12:57:57Z" level=debug msg="Process user information" gid=0 gids="1,2,3,4,6,10,11,20,26,27" name=root uid=0 username=root
time="2024-07-19T12:57:57Z" level=info msg="Authelia v4.38.9 is starting"
time="2024-07-19T12:57:57Z" level=info msg="Log severity set to debug"
time="2024-07-19T12:57:57Z" level=info msg="Storage schema is being checked for updates"
time="2024-07-19T12:57:57Z" level=info msg="Storage schema is already up to date"
time="2024-07-19T12:57:57Z" level=debug msg="LDAP Supported OIDs. Control Types: none. Extensions: 1.3.6.1.4.1.4203.1.11.1"
time="2024-07-19T12:57:58Z" level=debug msg="Create Server Service (metrics) skipped"
time="2024-07-19T12:57:58Z" level=info msg="Startup complete"
time="2024-07-19T12:57:58Z" level=info msg="Listening for non-TLS connections on '[::]:9091' path '/'" server=main service=server
time="2024-07-19T12:59:45Z" level=error msg="Request timeout occurred while handling request from client." error="read tcp 192.168.30.22:9091->192.168.30.22:33260: i/o timeout" method=GET path=/ remote_ip=192.168.30.22 stack="github.com/authelia/authelia/v4/internal/server/handlers.go:86 CreateDefaultServer.handleError.func2\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2870             (*Server).writeErrorResponse\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2308             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:224          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:196          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1695                                       goexit" status_code=408
time="2024-07-19T12:59:45Z" level=error msg="Request timeout occurred while handling request from client." error="read tcp 192.168.30.22:9091->192.168.30.22:33262: i/o timeout" method=GET path=/ remote_ip=192.168.30.22 stack="github.com/authelia/authelia/v4/internal/server/handlers.go:86 CreateDefaultServer.handleError.func2\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2870             (*Server).writeErrorResponse\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2308             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:224          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:196          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1695                                       goexit" status_code=408
time="2024-07-19T12:59:52Z" level=debug msg="Mark 1FA authentication attempt made by user 'myusername'" method=POST path=/api/firstfactor remote_ip="myclient_ipv6"
time="2024-07-19T12:59:52Z" level=debug msg="Successful 1FA authentication attempt made by user 'myusername'" method=POST path=/api/firstfactor remote_ip="myclient_ipv6"
time="2024-07-19T13:00:04Z" level=error msg="Request timeout occurred while handling request from client." error="read tcp 192.168.30.22:9091->192.168.30.22:45122: i/o timeout" method=GET path=/ remote_ip=192.168.30.22 stack="github.com/authelia/authelia/v4/internal/server/handlers.go:86 CreateDefaultServer.handleError.func2\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2870             (*Server).writeErrorResponse\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2308             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:224          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:196          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1695                                       goexit" status_code=408
time="2024-07-19T13:00:04Z" level=error msg="Request timeout occurred while handling request from client." error="read tcp 192.168.30.22:9091->192.168.30.22:45110: i/o timeout" method=GET path=/ remote_ip=192.168.30.22 stack="github.com/authelia/authelia/v4/internal/server/handlers.go:86 CreateDefaultServer.handleError.func2\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2870             (*Server).writeErrorResponse\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2308             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:224          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:196          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1695                                       goexit" status_code=408
time="2024-07-19T13:00:04Z" level=error msg="Request timeout occurred while handling request from client." error="read tcp 192.168.30.22:9091->192.168.30.22:45136: i/o timeout" method=GET path=/ remote_ip=192.168.30.22 stack="github.com/authelia/authelia/v4/internal/server/handlers.go:86 CreateDefaultServer.handleError.func2\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2870             (*Server).writeErrorResponse\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2308             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:224          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:196          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1695                                       goexit" status_code=408
time="2024-07-19T13:00:06Z" level=error msg="Error occurred retrieving TOTP configuration for user 'myusername': error occurred retrieving the configuration from the storage backend" error="no TOTP configuration for user" method=GET path=/api/secondfactor/totp remote_ip="myclient_ipv6" stack="github.com/authelia/authelia/v4/internal/handlers/handler_sign_totp.go:44 TimeBasedOneTimePasswordGET\ngithub.com/authelia/authelia/v4/internal/middlewares/require_auth.go:19   Require1FA.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/bridge.go:54         handleRouter.(*BridgeBuilder).Build.func6.1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:65        SecurityHeadersCSPNone.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:91        SecurityHeadersNoStore.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:30        SecurityHeadersBase.func1\ngithub.com/fasthttp/router@v1.5.1/router.go:441                           (*Router).Handler\ngithub.com/authelia/authelia/v4/internal/middlewares/log_request.go:14    handleRouter.LogRequest.func40\ngithub.com/authelia/authelia/v4/internal/middlewares/errors.go:38         RecoverPanic.func1\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2379                        (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:224                     (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:196                     (*workerPool).getCh.func1\nruntime/asm_amd64.s:1695                                                  goexit"
time="2024-07-19T13:06:58Z" level=error msg="Error occurred retrieving TOTP configuration for user 'myusername': error occurred retrieving the configuration from the storage backend" error="no TOTP configuration for user" method=GET path=/api/secondfactor/totp remote_ip="myclient_ipv6" stack="github.com/authelia/authelia/v4/internal/handlers/handler_sign_totp.go:44 TimeBasedOneTimePasswordGET\ngithub.com/authelia/authelia/v4/internal/middlewares/require_auth.go:19   Require1FA.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/bridge.go:54         handleRouter.(*BridgeBuilder).Build.func6.1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:65        SecurityHeadersCSPNone.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:91        SecurityHeadersNoStore.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:30        SecurityHeadersBase.func1\ngithub.com/fasthttp/router@v1.5.1/router.go:441                           (*Router).Handler\ngithub.com/authelia/authelia/v4/internal/middlewares/log_request.go:14    handleRouter.LogRequest.func40\ngithub.com/authelia/authelia/v4/internal/middlewares/errors.go:38         RecoverPanic.func1\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2379                        (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:224                     (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:196                     (*workerPool).getCh.func1\nruntime/asm_amd64.s:1695                                                  goexit"
time="2024-07-19T13:07:01Z" level=error msg="Error occurred retrieving TOTP configuration for user 'myusername': error occurred retrieving the configuration from the storage backend" error="no TOTP configuration for user" method=GET path=/api/secondfactor/totp remote_ip="myclient_ipv6" stack="github.com/authelia/authelia/v4/internal/handlers/handler_sign_totp.go:44 TimeBasedOneTimePasswordGET\ngithub.com/authelia/authelia/v4/internal/middlewares/require_auth.go:19   Require1FA.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/bridge.go:54         handleRouter.(*BridgeBuilder).Build.func6.1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:65        SecurityHeadersCSPNone.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:91        SecurityHeadersNoStore.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:30        SecurityHeadersBase.func1\ngithub.com/fasthttp/router@v1.5.1/router.go:441                           (*Router).Handler\ngithub.com/authelia/authelia/v4/internal/middlewares/log_request.go:14    handleRouter.LogRequest.func40\ngithub.com/authelia/authelia/v4/internal/middlewares/errors.go:38         RecoverPanic.func1\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2379                        (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:224                     (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:196                     (*workerPool).getCh.func1\nruntime/asm_amd64.s:1695                                                  goexit"
time="2024-07-19T13:07:04Z" level=error msg="Request timeout occurred while handling request from client." error="read tcp 192.168.30.22:9091->192.168.30.22:35510: i/o timeout" method=GET path=/ remote_ip=192.168.30.22 stack="github.com/authelia/authelia/v4/internal/server/handlers.go:86 CreateDefaultServer.handleError.func2\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2870             (*Server).writeErrorResponse\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2308             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:224          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:196          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1695                                       goexit" status_code=408
time="2024-07-19T13:07:04Z" level=error msg="Request timeout occurred while handling request from client." error="read tcp 192.168.30.22:9091->192.168.30.22:35518: i/o timeout" method=GET path=/ remote_ip=192.168.30.22 stack="github.com/authelia/authelia/v4/internal/server/handlers.go:86 CreateDefaultServer.handleError.func2\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2870             (*Server).writeErrorResponse\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2308             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:224          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:196          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1695                                       goexit" status_code=408
time="2024-07-19T13:07:04Z" level=error msg="Request timeout occurred while handling request from client." error="read tcp 192.168.30.22:9091->192.168.30.22:35534: i/o timeout" method=GET path=/ remote_ip=192.168.30.22 stack="github.com/authelia/authelia/v4/internal/server/handlers.go:86 CreateDefaultServer.handleError.func2\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2870             (*Server).writeErrorResponse\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2308             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:224          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:196          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1695                                       goexit" status_code=408
time="2024-07-19T13:07:04Z" level=error msg="Request timeout occurred while handling request from client." error="read tcp 192.168.30.22:9091->192.168.30.22:35548: i/o timeout" method=GET path=/ remote_ip=192.168.30.22 stack="github.com/authelia/authelia/v4/internal/server/handlers.go:86 CreateDefaultServer.handleError.func2\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2870             (*Server).writeErrorResponse\ngithub.com/valyala/fasthttp@v1.54.0/server.go:2308             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:224          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.54.0/workerpool.go:196          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1695                                       goexit" status_code=408

``` 
- Network and Console tabs both show 2 errors from 404 responses: `/api/secondfactor/totp` and `/api/secondfactor/webauthn/credentials` (firefox & chromium)
![2024-07-19_14:11:35](https://github.com/user-attachments/assets/e80eff8f-c744-483a-a965-4680619db6ba)
![2024-07-19_14:37:21](https://github.com/user-attachments/assets/80350bb1-8e1b-4b9b-ba1a-7a3f0c8eafaa)
- The console.error from console screenshot:
```json
{
    "message": "Request failed with status code 404",
    "name": "AxiosError",
    "stack": "AxiosError: Request failed with status code 404\n    at sS (https://auth.mydomain.com/static/js/index.R_EEPxQb.js:919:1042)\n    at XMLHttpRequest.f (https://auth.mydomain.com/static/js/index.R_EEPxQb.js:919:5886)\n    at Ll.request (https://auth.mydomain.com/static/js/index.R_EEPxQb.js:921:1952)\n    at async cj (https://auth.mydomain.com/static/js/index.R_EEPxQb.js:922:5626)\n    at async li (https://auth.mydomain.com/static/js/settings.router.5G8e5qOJ.js:29:36083)\n    at async https://auth.mydomain.com/static/js/index.R_EEPxQb.js:1079:726",
    "config": {
        "transitional": {
            "silentJSONParsing": true,
            "forcedJSONParsing": true,
            "clarifyTimeoutError": false
        },
        "adapter": [
            "xhr",
            "http",
            "fetch"
        ],
        "transformRequest": [
            null
        ],
        "transformResponse": [
            null
        ],
        "timeout": 0,
        "xsrfCookieName": "XSRF-TOKEN",
        "xsrfHeaderName": "X-XSRF-TOKEN",
        "maxContentLength": -1,
        "maxBodyLength": -1,
        "env": {},
        "headers": {
            "Accept": "application/json, text/plain, */*"
        },
        "method": "get",
        "url": "/api/secondfactor/webauthn/credentials"
    },
    "code": "ERR_BAD_REQUEST",
    "status": 404
}

``` 
I don't know if it is relevant but my `webauthn` section in the config has `disable: true`

**Reply by james-d-elliott** [reply]
I can't replicate this. I can replicate the unfriendly errors (which I am working on a fix for) but the TOTP registration panel still shows.

1. Are you using HTTPS in the browser?
2. After the problem occurs if you close the settings and refresh the page is the user still logged in?
3. What happens if you enable WebAuthn again?
4. What proxy are you using? Are you setting the X-Forwarded-* headers?

**Reply by mouchabeurre** [reply]
1. Yes, using HTTPS
2. Yes, the user is still logged in after closing the settings after the problem occurred (no page refresh needed, but also logged in if I do refresh)
3. When setting `disable: false` for WebAuthn the overall issue remains, but now the 404 response from `/api/secondfactor/webauthn/credentials` is 200 (with payload `{"status":"OK","data":null}`). Also: no error snackbar and no error in web console.
4. I'm using caddy without any X-Forwarded-* headers set, with the following basic config :
```Caddyfile
(secure) {
    forward_auth {args[0]} :9091 {
        uri /api/authz/forward-auth
        copy_headers Remote-User Remote-Groups Remote-Email Remote-Name
    }
}
*.mydomain.com {
    @ping host ping.mydomain.com
    handle @ping {
        import secure *
        respond "pong admin"
    }
    @authelia host auth.mydomain.com
    handle @authelia {
        reverse_proxy :9091
    }
}
``` 
_the following are uneducated guesses / I'm not familiar with the inner workings of Authelia_
Since my setup is working as expected minus the TOTP enrollment, I would think that the issue might lie on the storage side of Authelia? I'm using the sqlite storage backend, and I found the `totp_configurations` table to be empty, which is probably expected since no user has ever registered a device. I assume this is the place where the server is trying to retrieve the TOTP configuration?
- If it is, is it supposed to be empty? (I tried deleting my db to make Authelia generate a new one, the table stays empty after my user first login).
- If it is supposed to be empty, why would the server **expect** a record for the user on his first device enrollment?

**Reply by james-d-elliott** [reply]
The user would not be expected to have a value before they register and would not have one. The error realistically should be ignored because it's an indication that the credential needs to be registered but is being logged. I see the same logs but am presented with the option to register.

We actually test this specific scenario in every change made to Authelia and it's 100% reliable. I don't doubt there is an issue for you but the cause is not very obvious unless you're using an old version which doesn't seem to be the case. Only moonshot I can think of is a browser cache issue at this stage.

**Reply by mouchabeurre** [reply, marked as answer]
Good news, I think I figured out what is causing this issue:
There needs to be **at least one access control rule with a policy of `two_factor`** in the config for the TOTP setup to be shown at all in the web client.

- Having such a rule correctly greets an authenticated user on the `/2fa/one-time-password` route, allowing setting up the TOTP:

![Screenshot 2024-07-22 at 17-47-19 Login - Authelia](https://github.com/user-attachments/assets/b68f6f3f-43df-435f-a99e-befadda7cdf0)

- Not having such a rule greets an authenticated user on the `/authenticated` route, without any prompt for TOTP enrollment, even preventing TOTP setup when navigating to the 2FA settings:

![Screenshot 2024-07-22 at 17-46-19 Login - Authelia](https://github.com/user-attachments/assets/a7e87749-02e9-4a3d-a0aa-35ac99561b1e)

My mistake was to initially set all my routes to be protected with `one_factor`, thinking I could later add 2FA to my users and, only then, change my rules to `two_factor`. Don't ask me why I planned on doing this backwards!
Anyway, thanks for being so quick to help me troubleshoot my issue!
Perhaps explaining this behavior in the documentation on either [Home/Configuration/Second Factor/Time-based One-Time Password](https://www.authelia.com/configuration/second-factor/time-based-one-time-password/) or [Home/Configuration/Security/Access Control#two_factor](https://www.authelia.com/configuration/security/access-control/#two_factor) could help other users in the long run?

**Reply by james-d-elliott** [reply]
Glad you figured it out. I'll see what I can do to make this more intuitive from the UI, even though it is documented. I don't think adding it to multiple documentation pages is however the answer here. But I'll double check where it is and make sure it makes sense.

**Reply by winnie-sg** [reply]
> Good news, I think I figured out what is causing this issue: There needs to be **at least one access control rule with a policy of `two_factor`** in the config for the TOTP setup to be shown at all in the web client.

It was also my problem. I have spent a whole afternoon figuring out what's wrong with my configuration. Then I found your comment. Thank you so much.

**Reply by gene1wood** [reply]
I also encountered this.

This seems like it's not the right way to do this.

Even if no `access_control` stanza has a `policy` that *requires* [`two_factor`](https://www.authelia.com/configuration/security/access-control/#two_factor), users should still be able to enroll a two factor device and use it.

Just because the server side doesn't *require* two factor for something shouldn't mean that security minded users can't choose to enable two factor for their user account.

Or am I thinking about this wrong?

**Comment by cgkantidis** [comment]
I'm having the same issue, but the solution marked here doesn't work.
In my `configuration.yml` I have the following section:

```yml
access_control:
  default_policy: deny
  rules:
    - domain:
      - "auth.[mydomain].com"
      - "immich.[mydomain].com"
      policy: two_factor
```

I go to `auth.[mydomain].com` and signin with the username/password I have registered in `users_database.yml`.

I see the following screen:

<img width="568" height="680" alt="20250714_214515_red" src="https://github.com/user-attachments/assets/ac9fd35c-7bce-49d5-8e31-4bb195c56f60" />

and in the docker container log I see the following (I'm pasting the full log):

```
authelia  | time="2025-07-14T19:43:01Z" level=debug msg="Loaded Configuration Sources" files="[/config/configuration.yml]" filters="[]"
authelia  | time="2025-07-14T19:43:01Z" level=debug msg="Logging Initialized" fields.level=debug file= format= keep_stdout=false
authelia  | time="2025-07-14T19:43:01Z" level=debug msg="Process user information" gid=1000 uid=1000
authelia  | time="2025-07-14T19:43:01Z" level=info msg="Authelia v4.39.5 is starting"
authelia  | time="2025-07-14T19:43:01Z" level=info msg="Log severity set to debug"
authelia  | time="2025-07-14T19:43:01Z" level=debug msg="Registering OpenID Connect 1.0 client with client id 'immich' and policy 'two_factor'"
authelia  | time="2025-07-14T19:43:01Z" level=info msg="Storage schema is being checked for updates"
authelia  | time="2025-07-14T19:43:01Z" level=info msg="Storage schema is already up to date"
authelia  | time="2025-07-14T19:43:01Z" level=debug msg="webauthn-metadata provider: startup check skipped as it is disabled"
authelia  | time="2025-07-14T19:43:01Z" level=info msg="Startup complete"
authelia  | time="2025-07-14T19:43:01Z" level=info msg="Listening for non-TLS connections on '[::]:9091' path '/'" server=main service=server
authelia  | time="2025-07-14T19:45:02Z" level=debug msg="Mark 1FA authentication attempt made by user 'christos'" method=POST path=/api/firstfactor remote_ip=[ip]
authelia  | time="2025-07-14T19:45:02Z" level=debug msg="Successful 1FA authentication attempt made by user 'christos'" method=POST path=/api/firstfactor remote_ip=[ip]
```

Then, when I click on the `Register device` link, I get the following error in the docker logs:

```
authelia  | time="2025-07-14T19:47:39Z" level=error msg="Error occurred retrieving TOTP configuration for user 'christos': error occurred retrieving the configuration from the storage backend" error="no TOTP configuration for user" method=GET path=/api/secondfactor/totp remote_ip=[ip] stack="github.com/authelia/authelia/v4/internal/handlers/handler_sign_totp.go:43 TimeBasedOneTimePasswordGET\ngithub.com/authelia/authelia/v4/internal/middlewares/require_auth.go:19   Require1FA.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/bridge.go:66         handlerMain.(*BridgeBuilder).Build.func8.1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:65        SecurityHeadersCSPNone.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:105       SecurityHeadersNoStore.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:30        SecurityHeadersBase.func1\ngithub.com/fasthttp/router@v1.5.4/router.go:441                           (*Router).Handler\ngithub.com/authelia/authelia/v4/internal/middlewares/log_request.go:14    handlerMain.LogRequest.func30\ngithub.com/authelia/authelia/v4/internal/middlewares/errors.go:38         RecoverPanic.func1\ngithub.com/valyala/fasthttp@v1.63.0/server.go:2455                        (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.63.0/workerpool.go:225                     (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.63.0/workerpool.go:197                     (*workerPool).getCh.func1\nruntime/asm_amd64.s:1700                                                  goexit"
```

Here is also the `totp` section of my `configuration.yml`:

```
totp:
  disable: false
  issuer: 'authelia.com'
  algorithm: 'sha1'
  digits: 6
  period: 30
  skew: 1
  secret_size: 32
  allowed_algorithms:
    - 'SHA1'
  allowed_digits:
    - 6
  allowed_periods:
    - 30
  disable_reuse_security_policy: false
```

Anyone know what gives?

Authelia works great with 1 factor, but I really want to be able to setup the 2FA as well.

If needed, I can open a new issue, but this seemed so close to my issue that I'm not sure it's warranted.

Thanks in advance! Any help is appreciated!

**Reply by Crowley723** [reply]
That error is expected when you don't have any configured TOTP credentials. 

Are you prompted for an otp for session elevation?


You may want to open a fresh discussion so we aren't pinging people.
